src/nectarchain/calibration/NectarGain/SPEfit/utils.py

```diff
@@ -0,0 +1,281 @@
+import logging
+import math
+
+import numpy as np
+from iminuit import Minuit
+from scipy import interpolate, signal
+from scipy.special import gammainc
+from scipy.stats import norm
+
+logging.basicConfig(format='%(asctime)s %(name)s %(levelname)s %(message)s')
+log = logging.getLogger(__name__)
+log.handlers = logging.getLogger('__main__').handlers
+
+__all__ = ['UtilsMinuit','Multiprocessing']
+from .parameters import Parameters
+
+class Multiprocessing() : 
+    @staticmethod
+    def custom_error_callback(error):
+        log.error(f'Got an error: {error}')
+        log.error(error,exc_info=True)
+
+class UtilsMinuit() :
+    @staticmethod
+    def make_minuit_par_kwargs(parameters : Parameters):
+        """Create *Parameter Keyword Arguments* for the `Minuit` constructor.
+
+        updated for Minuit >2.0
+        """
+        names = parameters.parnames
+        kwargs = {"names": names,"values" : {}}
+
+        for parameter in parameters.parameters:
+            kwargs["values"][parameter.name] = parameter.value
+            min_ = None if np.isnan(parameter.min) else parameter.min
+            max_ = None if np.isnan(parameter.max) else parameter.max
+            error = 0.1 if np.isnan(parameter.error) else parameter.error
+            kwargs[f"limit_{parameter.name}"] = (min_, max_)
+            kwargs[f"error_{parameter.name}"] = error
+        return kwargs
+    
+    @staticmethod
+    def set_minuit_parameters_limits_and_errors(m : Minuit,parameters : dict) :
+        """function to set minuit parameter limits and errors with Minuit >2.0
+
+        Args:
+            m (Minuit): a Minuit instance
+            parameters (dict): dict containing parameters names, limits errors and values
+        """
+        for name in parameters["names"] :
+            m.limits[name] = parameters[f"limit_{name}"]
+            m.errors[name] = parameters[f"error_{name}"]
+
+
+# Usefull fucntions for the fit
+def gaussian(x, mu, sig):
+    #return (1./(sig*np.sqrt(2*math.pi)))*np.exp(-np.power(x - mu, 2.) / (2 * np.power(sig, 2.)))
+    return norm.pdf(x,loc = mu,scale = sig)
+
+def weight_gaussian(x,N, mu, sig) : 
+    return N * gaussian(x, mu, sig)
+
+def doubleGauss(x,sig1,mu2,sig2,p):
+    return p *2 *gaussian(x, 0, sig1) + (1-p) * gaussian(x, mu2, sig2)
+
+def PMax(r):
+    """p_{max} in equation 6 in Caroff et al. (2019)
+
+    Args:
+        res (float): SPE resolution
+
+    Returns:
+        float : p_{max}
+    """
+    pmax = np.pi*r**2/(np.pi*r**2 + np.pi - 2*r**2 - 2)
```

Comment:
In the equation 6 of the paper, it is `-2res` in the term below the divison bar, not `-2 r **2`

Reply:
Indeed, many thanks, but before to modify I will check if it is not again a typo in the article.

Reply:
OK. If it's a typo from the paper, please mention it in the documentation of the function.

Reply:
@guillaumegrolleron : Have you been able to check if this was a typo ?
Once it's done, I think it will be good to go.

Reply:
I have tried to derive equation of the paper to check, but I didn't have yet the pmax expression, and I detected other typo in equation 4, I am on it and I will do it as fast as possible.

Reply:
I derived expression of p_max (equation 6) and indeed there is a typo in the paper, the correct expression is with -2res^2 and not -2res.

Reply:
Hi @guillaumegrolleron ,
So the code is OK, thanks a lot for all these checks!